files/virt_volume.sh

```diff
@@ -17,82 +17,119 @@
     exit 1
 fi
 
-NAME=$1
-POOL=$2
-CAPACITY=$3
-FORMAT=$4
-IMAGE=$5
+while getopts ":n:p:c:f:i:b:" opt; do
+    case ${opt} in
+        n) NAME=$OPTARG;;
+        p) POOL=$OPTARG;;
+        c) CAPACITY=$OPTARG;;
+        f) FORMAT=$OPTARG;;
+        i) IMAGE=$OPTARG;;
+        b) BACKING_IMAGE=$OPTARG;;
+        \?)
+            echo "Invalid option: -$OPTARG" >&2
+            exit 1
+            ;;
+        :)
+            echo "Option -$OPTARG requires an argument." >&2
+            exit 1
+            ;;
+    esac
+done
+
+if [[ -n $IMAGE && -n $BACKING_IMAGE ]]; then
+  echo "Options -i and -b are mutually exclusive." >&2
+  exit 1
+fi
 
 # Check whether a volume with this name exists.
-output=$(virsh vol-info --pool $POOL --vol $NAME 2>&1)
+output=$(virsh vol-info --pool "$POOL" --vol "$NAME" 2>&1)
 result=$?
 if [[ $result -eq 0 ]]; then
     echo '{"changed": false}'
     exit 0
 elif ! echo "$output" | grep 'Storage volume not found' >/dev/null 2>&1; then
-    echo "Unexpected error while getting volume info"
+    echo "Unexpected error while getting volume info" >&2
     echo "$output"
     exit $result
 fi
 
 # Create the volume.
-output=$(virsh vol-create-as --pool $POOL --name $NAME --capacity $CAPACITY --format $FORMAT 2>&1)
-result=$?
+if [[ -n $BACKING_IMAGE ]]; then
+    if [[ "$FORMAT" -ne "qcow2" ]]; then
+        echo "qcow2 format assumed for backing images, but $FORMAT format was supplied."
+        exit 1
+    fi
+    output=$(virsh vol-create-as --pool "$POOL" --name "$NAME" --capacity "$CAPACITY" --format "$FORMAT" --backing-vol "$BACKING_IMAGE" --backing-vol-format "$FORMAT" 2>&1)
+    result=$?
+else
+    output=$(virsh vol-create-as --pool "$POOL" --name "$NAME" --capacity "$CAPACITY" --format "$FORMAT" 2>&1)
+    result=$?
+fi
 if [[ $result -ne 0 ]]; then
     echo "Failed to create volume"
     echo "$output"
     exit $result
 fi
 
 # Determine the path to the volume file.
-output=$(virsh vol-key --pool $POOL --vol $NAME 2>&1)
+output=$(virsh vol-key --pool "$POOL" --vol "$NAME" 2>&1)
 result=$?
 if [[ $result -ne 0 ]]; then
     echo "Failed to get volume file path"
     echo "$output"
-    virsh vol-delete --pool $POOL --vol $NAME
+    virsh vol-delete --pool "$POOL" --vol "$NAME"
     exit $result
 fi
 
 # Change the ownership of the volume to VOLUME_OWNER:VOLUME_GROUP if
 # these environmental variables are defined. Without doing this libvirt
 # cannot access the volume on RedHat based GNU/Linux distributions.
-existing_owner="$(stat --format '%U' "$output")"
-existing_group="$(stat --format '%G' "$output")"
-new_owner="${VOLUME_OWNER:-$existing_owner}"
-new_group="${VOLUME_GROUP:-$existing_group}"
-output=$(chown "$new_owner":"$new_group" $output 2>1)
-result=$?
-if [[ $result -ne 0 ]]; then
-    echo "Failed to change ownership of the volume to $new_owner:$new_group"
-    echo "$output"
-    virsh vol-delete --pool $POOL --vol $NAME
-    exit $result
+if [[ -z "$VOLUME_OWNER" || -z "$VOLUME_GROUP" ]]; then
```

Comment:
Should these be `-n`?

Reply:
Also, what method are you using to unset these environment variables?

Reply:
> Also, what method are you using to unset these environment variables?

Didn't unset them. The next line's test, `[[ -f "$output" || -d "$output" ]]`, only attempts to adjust permissions if the volume is file or directory based, since I noticed trying to change permissions on logical volumes (LVM) isn't a good idea if the storage pool is a volume group.`

Admittedly, that was a bit of a hack at avoiding a problem when `existing_owner="$(stat --format '%U' "$output")"` and related lines tried to do things with permissions on logical volumes. It probably worked okay for raw or qcow2 image files, but caused issues with LVM based pools for me.

I'm not sure if there are other storage volume types besides files and directories where it makes sense to force permissions.

FYI, at least on RHEL7 for the system users, libvirt goes and directly adjusts / manages for image files.

Reply:
Maybe a negative test to skip when a block device is the volume is another way, e.g. `! [[ -b "$output" ]]`

Reply:
I think it would only work for LVM backed volumes currently because the `-z` now always skips this chunk of code.

So for LVM, the output of `virsh vol-key` is the block device? In that case, I imagine `-b` should work.

Reply:
Okay, so testing for the existence of the env vars (with bad logic!) is unnecessary because I notice the code already defaults to trying to set to the existing owner. If I kept the check for the env vars, then should've used `-n` instead of `-z` test (mixed them up), but can actually strip it out.

Reply:
So kept `[[ -f "$output" || -d "$output" ]]`, but otherwise reverted the bad/unnecessary check for if VOLUME_GROUP or  VOLUME_OWNER are set.